crates/block-producer/src/mempool/dependency_graph.rs

```diff
@@ -0,0 +1,1000 @@
+            to_remove.extend(unvisited_children);
+
+            // Inform parents that this child no longer exists.
+            let parents = self.parents.remove(&key).unwrap_or_default();
+            for parent in parents {
+                if let Some(parent) = self.children.get_mut(&parent) {
+                    parent.remove(&key);
+                }
+            }
+        }
+
+        Ok(removed)
+    }
+
+    /// Adds the node to the `roots` list _IFF_ all of its parents are processed.
+    ///
+    /// # SAFETY
+    ///
+    /// This method assumes the node exists. Caller is responsible for ensuring this is true.
+    fn try_make_root(&mut self, key: K) {
+        if self.pending.contains(&key) {
+            return;
+        }
+        debug_assert!(
+            self.vertices.contains_key(&key),
+            "Potential root {key} must exist in the graph"
+        );
+        debug_assert!(
+            !self.processed.contains(&key),
+            "Potential root {key} cannot already be processed"
+        );
+
+        let all_parents_processed = self
+            .parents
+            .get(&key)
+            .into_iter()
+            .flatten()
+            .all(|parent| self.processed.contains(parent));
+
+        if all_parents_processed {
+            self.roots.insert(key);
+        }
+    }
+
+    /// Returns the set of nodes that are ready for processing.
+    ///
+    /// Nodes can be selected from here and marked as processed using [`Self::process_root`].
+    pub fn roots(&self) -> &BTreeSet<K> {
+        &self.roots
+    }
+
+    /// Marks a root node as processed, removing it from the roots list.
+    ///
+    /// The node's children are [evaluated](Self::try_make_root) as possible roots.
+    ///
+    /// # Error
+    ///
+    /// Errors if the node is not in the roots list.
+    ///
+    /// This method is atomic.
+    pub fn process_root(&mut self, key: K) -> Result<(), GraphError<K>> {
+        if !self.roots.remove(&key) {
+            return Err(GraphError::InvalidRootNode(key));
+        }
+
+        self.processed.insert(key);
+
+        self.children
+            .get(&key)
+            .cloned()
+            .unwrap_or_default()
+            .into_iter()
+            .for_each(|child| self.try_make_root(child));
+
+        Ok(())
+    }
+
+    /// Returns the value of a node.
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.vertices.get(key)
+    }
+
+    /// Returns the parents of the node, or [None] if the node does not exist.
+    pub fn parents(&self, key: &K) -> Option<&BTreeSet<K>> {
+        self.parents.get(key)
+    }
+
+    /// Returns true if the node exists, in either the pending or non-pending sets.
+    fn contains(&self, key: &K) -> bool {
+        self.pending.contains(key) || self.vertices.contains_key(key)
+    }
+}
+
+// TESTS
+// ================================================================================================
+
+#[cfg(test)]
+mod tests {
```

Comment:
nit: I would consider breaking this file up to move tests into a separate file - e.g., `src/mempool/graph/mod.rs` and `src/mempool/graph/tests.rs`.